extensions/sql-database-projects/src/models/project.ts

```diff
@@ -1106,6 +1089,56 @@ export class Project implements ISqlProject {
 
 		return firstPropertyElement.childNodes[0].data;
 	}
+
+	/**
+	 * Adds all folders in the path to the project and saves the project file, if provided path is under the project folder.
+	 * If path is outside the project folder, then no action is taken.
+	 *
+	 * @param relativeFolderPath Relative folder path to add folders from.
+	 * @returns Project entry for the last folder in the path, if path is under the project folder; otherwise `undefined`.
+	 */
+	private async ensureFolderItems(relativeFolderPath: string): Promise<FileProjectEntry | undefined> {
+		const absoluteFolderPath = path.join(this.projectFolderPath, relativeFolderPath);
+		const normalizedProjectFolderPath = path.normalize(this.projectFolderPath);
+
+		// Only add folders within the project folder. When adding files outside the project folder,
+		// they should be copied to the project root and there will be no additional folders to add.
+		if (!absoluteFolderPath.toUpperCase().startsWith(normalizedProjectFolderPath.toUpperCase())) {
+			return;
+		}
+
+		// If folder doesn't exist, create it
+		await fs.mkdir(absoluteFolderPath, { recursive: true });
+
+		// Add project file entries for all folders in the path.
+		// SSDT expects all folders to be explicitly listed in the project file, so we construct
+		// folder paths for all intermediate folders and ensure they are present in the project as well.
+		// We do not use `path.relative` here, because it may return '.' if paths are the same,
+		// but in our case we actually want an empty string, that will result in an empty segments
+		// array and nothing will be added.
+		const relativePath = utils.convertSlashesForSqlProj(absoluteFolderPath.substring(normalizedProjectFolderPath.length));
+		const pathSegments = utils.trimChars(relativePath, ' \\').split(constants.SqlProjPathSeparator);
+		let folderEntryPath = '';
+		let folderEntry: FileProjectEntry | undefined;
+
+		// Add folder items for all segments, including the requested folder itself
```

Comment:
Here's another caveat of routing everything through the same code - the trailing slash will now be enforced on all folder items - no matter whether you request to add `foo\bar` or `foo\bar\` it will always add `foo\bar\`. I think this is reasonable, given this is how SSDT "expects" them to be present.

If we don't want this normalization, then we will need to special case it in `addFolderItem` and this again becomes a helper method that only adds intermediate folders.